broadinstitute/single_cell_portal_core
# Restore customizing scatter plot colors (SCP-5112)

#### BACKGROUND & CHANGES
This fixes a bug with saving custom scatter plot colors where updates never persist, despite telling the user that the save was successful.  This was an issue likely introduced [here](https://github.com/broadinstitute/single_cell_portal_core/commit/64f7de70e0d2ab5ff347f453064538cdf3835256) as a part of work to improve the AnnData upload UX, where the `custom_color_updates` key was being de-nested and not passed to the server.  This key is a controlled value required for proper handling of color objects, and the omission caused all updates to be discarded by the server before saving.

#### MANUAL TESTING
1. Boot as normal and sign in and either a) turn off all caching, or b) start DelayedJob to process cache clearing jobs
2. Load any normal scatter plot and click the `Customize colors` link in the legend once it loads
3. Select the first entry and change the color, then click `Save colors`
4. Confirm the color saves and you see the success dialog
5. Refresh the page and confirm that the updated color is still shown
6. Click `Customize colors` again and then `Reset to defaults`, confirming the colors reset
7. Refresh the page again and confirm that the colors are back to the default palette
